### PR TITLE
added support for es7 decorator (a la angular2)

### PR DESCRIPTION
**Pull request by orizens**
fixes https://github.com/Microsoft/TypeScript-TmLanguage/issues/8 
followed from https://github.com/Microsoft/TypeScript-Sublime-Plugin/issues/476  
typescript with es7 descorators should like this now:  
screen shot 2016-06-03 at 2 36 27 pm

### CODE REVIEW

**Comment by mhegazy** [comment]
@anubmat  can you take a look.

**Comment by anubmat** [comment]
@orizens , Thank you for your contribution, However, there are two modifications I would like you to make before I can review these changes:
First, I see you have made changes to the pList file - "TypeScript.tmLanguage". There is a YAML file "TypeScript.YAML-tmLanguage" that can be converted to this pList file. It is easy to make changes to the YAML and convert it to pList. We do not want the YAML to be stale, so they must remain in sync. I can help you out here if you need help with this. 
Secondly, there is another YAML file - "TypeScriptReact.YAML-tmLanguage" that also converts to pList file by the name: "TypeScriptReact.tmLanguage". Corresponding changes need be made to this file so that TypeScript and TypeScriptReact remain in sync.
So in total, there must be 4 files changed - 2 YAMLs and 2 pLists

**Comment by orizens** [comment]
hi @anubmat 
I can update the yaml - no probs. 
Regrading conversion - sublime can now read YAML with "sublime-syntax" extension. 
Which tool you're using to convert to tmLangauge?

**Comment by anubmat** [comment]
Hi @orizens , You can use sublime text's package dev after installing [package control](https://github.com/SublimeText/PackageDev#getting-started). Here is a link to the repo and details on how to use it - https://github.com/SublimeText/PackageDev. It will generate a tmLanguage (pList) from the YAML.

**Comment by orizens** [comment]
hi @anubmat 
I updated the yaml file and generated a tmLanguage file.

**Comment by anubmat** [comment]
Hi @orizens , can you please check why some tests are failing, by running npm test from tests folder, maybe you have to fix some baselines. Also as mentioned earlier, can you also make changes to TypeScriptReact.YAML-tmLanguage and its corresponding tmLanguage file. Also it will be great if you can add some testcases.

**Comment by orizens** [comment]
hi @anubmat 
i updated the TypescriptReactYML. 
however - i can't figure the tests failures.

File: TypeScript.YAML-tmLanguage [hunk 1]
```diff
@@ -413,6 +412,187 @@ repository:
     - include: '#function-call'
     - include: '#case-clause'
     - include: '#control-statement'
+    - include: '#es7-decorators'
+    - include: '#literal-class'
+    - include: '#literal-constructor'
+  
+  literal-constructor:
+    patterns:
+    - name: meta.instance.constructor.js
```
Comment (anubmat) [inline TypeScript.YAML-tmLanguage]: names in this files go as *.ts

**Comment by anubmat** [comment]
Refer to this[ link ](https://github.com/Microsoft/TypeScript-TmLanguage#tests) for details on how to do the testing locally. Some of your tests failed because you introduced new scopes, which is okay but you may have to go over these tests, find differences, compare results in test/generated folder against test/baselines folder and either update the files in test/baselines or improve some of your changes to pass existing baselines.

**Comment by orizens** [comment]
hi @anubmat 
sorry - I don't understand how the tests for this package works.

File: TypeScript.YAML-tmLanguage [hunk 2]
```diff
@@ -495,16 +675,16 @@ repository:
     - include: '#expression'
 
   expression-operator:
-    name: keyword.others.ts
-    match: \b(delete|in|instanceof|new|typeof|as|is|of)\b
+    name: keyword.operator.ts
```
Comment (anubmat) [inline TypeScript.YAML-tmLanguage]: Is there any particular reason why you changed the name of the scope?

Comment (orizens) [inline TypeScript.YAML-tmLanguage]: that was a while ago.
i think that i did it to color those keyword in a different color.

Comment (anubmat) [inline TypeScript.YAML-tmLanguage]: Ok, I see, we try our best to associate 'operator' with actual operators (mathematical or logical). So I'd suggest reverting this change.

**Comment by anubmat** [comment]
I am looking at the test failures, you are failing 14/33 tests. I am pointing out a few places, where changes can be made. 
Just for your reference here is the process to test:
from the folder 'TypeScript-TmLanguage\tests', run:
             `npm install`
             `npm test`
It will list failed cases. The test case files are present in 'TypeScript-TmLanguage\tests\cases'. In these test typescript files, '^^' is used to test the scope of the following word/code. 
Every testcase is run with npm tests and generates files in 'TypeScript-TmLanguage\tests\generated' folder. These are compared against existing files ( or baselines) in 'TypeScript-TmLanguage\tests\baselines' folder. 
The differences in these files is what leads to failing tests. You can always compare these files, and if you think the ones generated have better scopes, you can change the corresponding baseline and check-in the updated baseline. If on the other hand you feel your generated files give unacceptable scopes, you should update these grammar files to correct those scopes.

File: TypeScript.YAML-tmLanguage [hunk 3]
```diff
@@ -66,7 +65,7 @@ repository:
 
   object-declaration:
     name: meta.declaration.object.ts
-    begin: '\b(?:(export)\s+)?\b(?:(abstract)\s+)?\b(?<!\.)(class|interface)\b'
+    begin: '\b(?:\s+)?\b(?:(abstract)\s+)?\b(?<!\.)(class|interface)\b'
```
Comment (anubmat) [inline TypeScript.YAML-tmLanguage]: I am not sure, why would you want to look ahead for spaces - (?:\s+)

**Comment by orizens** [comment]
i'm not sure either. 
i think i'll start from scratch.

File: TypeScriptReact.YAML-tmLanguage [hunk 4]
```diff
@@ -415,7 +415,187 @@ repository:
     - include: '#function-call'
     - include: '#case-clause'
     - include: '#control-statement'
+    - include: '#es7-decorators'
```
Comment (DanielRosenwasser) [inline TypeScriptReact.YAML-tmLanguage]: This should just be called `decorator`

**Comment by DanielRosenwasser** [comment]
Hey @orizens! If you do send a follow-up PR, just a few comments based on what you've sent so far:

* Do not base this work off of other repos. It appears that you adapted the work from [the Babel TmLanguage](https://github.com/babel/babel-sublime/blob/548e8150ae01ac015718973668bacfc7262110c2/JavaScript%20(Babel).YAML-tmLanguage) file, where the license is currently not compatible and the adapted portions don't contain attribution. I don't think it's a good idea to try to "meld" the files together like this anyway.
* Keep versions and project names (like "ES7" and "flowtype") out of scope names.

Feel free to ping us if you have any questions.

**Comment by orizens** [comment]
hi @DanielRosenwasser 
notes your suggestions.
I opened a new pr - #201 
There were too many confusions with this one. 
With the new PR - I also added html syntax highlight in template strings - which is a much needed feature . 
You're right about adopting the es7 decorators from the babel package - i noted that on the new pr and will add it the yml as well. 
I'm not sure about licensing and using bable's MIT code in this project - that's currently beyond me. 
It took me few tries to make decorator highlighting work with typescript's yml - and i'm working with this for about 3 months now - without any side effects (before i opened this pr).

**Pull request closed without merging**